### DISCUSSION TITLE
Custom Cuda Kernels for Image Processing

### DISCUSSION THREAD

**Discussion by vade** [kornia/kornia#1695] [Questions]
Hello friends

Im a graphics dev looking to help my data science colleague out by implementing some fast, custom GPU augmentations.

I am curious how one would approach loading custom kernels to Kornia. Apologies, im unfamiliar with some of the existing PyTorch 'ways of doing things' - so if this is basic dont hesitate to point me to some docs I may not have seen.

Ive begun experimenting with CUDA-Python which is NVidias new and supposedly standard method of directly calling CUDA from Python. Is this interoperable with Kornia?

https://github.com/Synopsis/CUDA-Python-Augmentations/blob/main/Grayscale%20Cuda%20Kernel.ipynb

It might be nice to have a basic Kornia class that exposes a Kernel, methods to supply arguments / buffers, and that manages GPU memory life cycles. 

That would allow for dramatically fast experimentation in CUDA with new / novel augmentations by folks who are domain experts (graphics etc) without needing to understand the other nuances in the rest of the env.

Any advice or pointers would be greatly appreciated.

Thank you.

**Comment by edgarriba** [comment]
@vade this an awesome idea, check on this: https://pytorch.org/docs/stable/generated/torch.cuda.Stream.html

At a high level I could see something like

```python
img: Image = K.io.load_image(…)  CxHxW
Img = img.cuda()

kernel = "…" # string or path to .cu kernel
opts = CudaKernelConfig(kernel, img.stream()) # we might need to overload the Image class to get the stream 
fcn = K.core.cudaKernel(opts)

out = fcn(img)

```
Happy to iterate on this design. Which I believe that can go according to this attempt of having a generic Image class to handle different frameworks (not only pytorch).  https://github.com/kornia/kornia/pull/1562 I’m slowly polishing the design based on cross projects I’m working on right now specially from robotics perspective .

**Reply by vade** [reply]
Interesting. Would `fcn` allow for args to be calculated for the specific kernel / augmentation? I ask because ideally some of these would have procedurally driven parameterization to allow for some variability between runs. Could one align the CUDA args / types for the cuda kernel being evoked?

That above code looks pretty ideal tbh - very light weight. 

How can I help / what's the best next step?

**Reply by edgarriba** [reply]
as a first step I’d say encapsulate your code in a python class, then we can iterate to make it closer to my proposal. Something to look as a reference too is https://pytorch.org/tutorials/advanced/cpp_extension.html — but this will limit to pytorch

**Reply by vade** [reply]
Awesome. Ill work on this and iterate on it via the notebook code I published. Thanks @edgarriba for guidance and your input.
